modules/upscaler.py

```diff
@@ -57,7 +57,7 @@ def upscale(self, img: PIL.Image, scale, selected_model: str = None):
         dest_h = int((img.height * scale) // 8 * 8)
 
         for _ in range(3):
-            if img.width >= dest_w and img.height >= dest_h:
+            if img.width > dest_w and img.height > dest_h:
                 break
```

Comment:
Can you explain why this (and the other `if` at the end of the loop) are necessary? 🤔 As I see it, this implementation will do an extra upscale, just to downscale back down with lanczos outside the loop?

Reply:
The line at bottom was part of https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/30b1bcc64e67ad50c5d3af3a6fe1bd1e9553f34e (1x scale models work with this) . While the new one https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/4a666381bf98333ba4512db0f0033df5f6a08771  just moved the line to the top of the break the loop before upscale has a chance to running under the same conditions of the prior commit. This had the side effect of break 1x scale models, which would previously be run at which point the line at bottom would just return the output immediately?

I can only assume the lanczos downscale is intentional? Maybe it is what allows users to get custom output sizes out of 2x/3x/4x/8x upscaling models? 

AUTOMATIC1111 would likely need to chime in here if you believe the entire loop is wrong.

Reply:
Yes, as far as I can tell the Lanczos (down)scale is intentional, exactly to get fractional scales out of models that are generally just 2x/3x/4x. 

I think at least a code comment is needed to explain the logic to support 1x models – and in the case of an 1x model, it might be better to do the Lanczos upscaling first, then refine the scaled output with an 1x model, right?

Reply:
I think I get what you are saying about Lanczos, but do people actually willing scale the image with a 1x models selected rather then just use them in Extras, considering how they are implemented currently?

This really comes back to if you are going to be refactoring the scaling code, consider adding something to parse and record the model scale somewhere like cache.json after the model is loaded for the first time. Or maybe another option would be to dump them in a separate sub-folder where all models there are implicitly considered as 1x post-processing models and can be handled differently. 

Overall I think trying to further adapt this section to 1x models without actually knowing a 1x model is being used is impractical. I'll add a comment there though, explaining what is happening.